server/src/main/java/org/elasticsearch/common/io/stream/RecyclerBytesStreamOutput.java

```diff
@@ -0,0 +1,230 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.common.io.stream;
+
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.common.bytes.BytesArray;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.bytes.CompositeBytesReference;
+import org.elasticsearch.common.recycler.Recycler;
+import org.elasticsearch.common.util.PageCacheRecycler;
+import org.elasticsearch.core.Releasable;
+import org.elasticsearch.core.Releasables;
+import org.elasticsearch.transport.BytesRefRecycler;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.VarHandle;
+import java.nio.ByteOrder;
+import java.util.ArrayList;
+
+/**
+ * A @link {@link StreamOutput} that uses {@link Recycler.V<BytesRef>} to acquire pages of bytes, which
+ * avoids frequent reallocation &amp; copying of the internal data. When {@link #close()} is called,
+ * the bytes will be released.
+ */
+public class RecyclerBytesStreamOutput extends BytesStream implements Releasable {
+
+    static final VarHandle VH_BE_INT = MethodHandles.byteArrayViewVarHandle(int[].class, ByteOrder.BIG_ENDIAN);
+    static final VarHandle VH_BE_LONG = MethodHandles.byteArrayViewVarHandle(long[].class, ByteOrder.BIG_ENDIAN);
+
+    private final ArrayList<Recycler.V<BytesRef>> pages = new ArrayList<>();
+    private final Recycler<BytesRef> recycler;
+    private final int pageSize;
+    private int pageIndex = -1;
+    private int currentCapacity = 0;
+    private int currentPageOffset;
+
+    public RecyclerBytesStreamOutput() {
```

Comment:
I wonder if we should drop this constructor for now since this StreamOutput doesn't work efficiently if we're not effectively recycling bytes, and explicitly pass `PageCacheRecycler.NON_RECYCLING_INSTANCE` in tests.